ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.engine.KsqlEngine;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.confluent.ksql.util.QueryMetadata;
+import java.net.URI;
+import java.net.URL;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.state.HostInfo;
+import org.apache.kafka.streams.state.StreamsMetadata;
+
+public final class HeartbeatHandler {
+
+  private final KsqlEngine engine;
+  private final ServiceContext serviceContext;
+  private final ConfigurationParameters configs;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> receivedHeartbeats;
+  private final ConcurrentHashMap<String, Boolean> aliveHosts;
+  private final ConcurrentHashMap<URI, Boolean> remoteHostsURI;
+  private final ConcurrentHashMap<HostInfo, Boolean> remoteHostsInfo;
+  private final HostInfo localHostInfo;
+  private final URL localURL;
+
+
+  public static HeartbeatHandler.Builder builder() {
+    return new HeartbeatHandler.Builder();
+  }
+
+  private HeartbeatHandler(KsqlEngine engine,
+                           ServiceContext serviceContext,
+                           URI localServerURI,
+                           ConfigurationParameters configs) {
+
+    this.engine = requireNonNull(engine, "engine");
+    this.serviceContext = requireNonNull(serviceContext, "serviceContext");
+    this.configs = requireNonNull(configs, "configuration parameters");
+    this.scheduledExecutorService = Executors.newScheduledThreadPool(configs.threadPoolSize);
+    this.receivedHeartbeats = new ConcurrentHashMap<>();
+    this.aliveHosts = new ConcurrentHashMap<>();
+    this.remoteHostsURI = new ConcurrentHashMap<>();
+    this.remoteHostsInfo = new ConcurrentHashMap<>();
+    this.localHostInfo = new HostInfo(localServerURI.getHost(), localServerURI.getPort());
+    try {
+      this.localURL = localServerURI.toURL();
+    } catch (final Exception e) {
+      throw new IllegalStateException("Failed to convert local host URI to URL."
+                                          + " local host URL: " + localServerURI);
+    }
+    this.aliveHosts.putIfAbsent(localHostInfo.toString(), true);
+  }
+
+  void scheduleHeartbeatRelatedTasks() {
+    SendHeartbeatTask sendTask = new SendHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(sendTask, 0, configs.heartbeatSendIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    CheckHeartbeatTask processTask = new CheckHeartbeatTask();
```

Comment:
if one of these dies, you probably want the other to die too and error out violently - `AbstractScheduledService` would help you there

Reply:
I assume you are talking about the guava class? +1 as a nicer abstraction anyway.

But in this case, I don't think you would want to fail them right?  for e.g if sending heartbeats fail, why would we want to not check heartbeats..  Nonetheless @vpapavas lets ensure the `try-catch`in all of these tasks, catch `Throwable` and thus make sure, it will not error out. ?

Reply:
Yeah I was considering the guava class - I get the independence of the tasks, but I need to reason it out a bit more about what happens if the threads die... I'll comment what I think about above in my original comment